Source/Libraries/GSF.Security/OIDCSecurityProvider.cs

```diff
@@ -0,0 +1,582 @@
+
+#pragma warning disable S2068
+
+namespace GSF.Security
+{
+    /// <summary>
+    /// Represents an <see cref="ISecurityProvider"/> that uses openID Connect
+    /// </summary>
+    /// <example>
+    /// Required config file entries (automatically added):
+    /// <code>
+    /// <![CDATA[
+    /// <?xml version="1.0"?>
+    /// <configuration>
+    ///   <configSections>
+    ///     <section name="categorizedSettings" type="GSF.Configuration.CategorizedSettingsSection, GSF.Core" />
+    ///   </configSections>
+    ///   <categorizedSettings>
+    ///     <securityProvider>
+    ///       <add name="ProviderType" value="GSF.Security.OIDCSecurityProvider, GSF.Security" description="The type to be used for enforcing security."
+    ///         encrypted="false" />
+    ///       <add name="ClientID" value="xxxx-xxxx-xxxx" description="Defines the ClientID as required per OpenID Connect Standard." encrypted="false" />
+    ///       <add name="Scope" value="user" description="Defines the Scope as required per OpenID Connect Standard." encrypted="false" />
+    ///       <add name="AuthorizationEndpoint" value="user" description="Defines the Endpoint to redirect the user for Authorization." encrypted="false" />
+    ///       <add name="RedirectURI" value="https://localhost:8986/" description="Defines the URI the User get's redirected to after signing in." encrypted="false" />
+    ///       <add name="ClientSecret" value="sssss-ssssss-sssss" description="Defines the Client Secret to encrypt User Information." encrypted="false" />
+    ///       <add name="TokenEndpoint" value="user" description="Defines the Endpoint to get the User Token from." encrypted="false" />
+    ///     </securityProvider>
+    ///   </categorizedSettings>
+    /// </configuration>
+    /// ]]>
+    /// </code>
+    /// </example>
+
+    public class OIDCSecurityProvider : SecurityProviderBase
+    {
+        #region [ Members ]
+
+        // Nested Types
+        private class TokenResponse
+        {
+            public string access_token { get; set; }
+            public string id_token { get; set; }
+        }
+
+        // Constants
+        private const string ResponseType = "code";                          // Response Type the SecurityProvider expects
+        private const int NonceSlidingExpiration = 600;                      // Time in Seconds that a Nonce is valid for. This prevents replay attacks
+       
+        /// <summary>
+        /// Defines the provider ID for the <see cref="AdoSecurityProvider"/>.
+        /// </summary>
+        public new const int ProviderID = 3;
+
+        #endregion
+
+        #region [ Constructors ]
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AdoSecurityProvider"/> class.
+        /// </summary>
+        /// <param name="username">Name that uniquely identifies the user.</param>
+        public OIDCSecurityProvider(string username)
+            : this(username, true, false, false)
```

Comment:
This provider's `RefreshData()` method simply returns false so I don't think `CanRefreshData` should be true.

```suggestion
            : this(username, false, false, false)
```